cpp/src/plane_extraction.cpp

```diff
@@ -0,0 +1,569 @@
+     {"depthSigmaMargin", "10"},
+     {"minPtsPerCell", "3"},
+     {"depthDiscontinuityThreshold", "160"},
+     {"maxNumberDepthDiscontinuity", "1"}}};
+
+CAPE::CAPE(int32_t image_height, int32_t image_width, config::Config config)
+    : _config(config),
+      _nr_horizontal_cells(image_width / config.getInt("patchSize")),
+      _nr_vertical_cells(image_height / config.getInt("patchSize")),
+      _nr_total_cells(_nr_horizontal_cells * _nr_vertical_cells),
+      _nr_pts_per_cell(pow(config.getInt("patchSize"), 2)),
+      _image_height(image_height),
+      _image_width(image_width),
+      _cell_grid(_nr_total_cells, nullptr),
+      _grid_plane_seg_map(_nr_vertical_cells, _nr_horizontal_cells, 0),
+      _grid_plane_seg_map_eroded(_nr_vertical_cells, _nr_horizontal_cells,
+                                 uchar(0)),
+      _seg_map_stacked(image_height * image_width, 0) {}
+
+cv::Mat CAPE::process(Eigen::MatrixXf const& pcd_array) {
+  // 0. Stack array by cell
+  Eigen::MatrixXf organized_array(pcd_array.rows(), pcd_array.cols());
+  organizeByCell(pcd_array, &organized_array);
+  // 1. Planar cell fitting
+  std::bitset<BITSET_SIZE> planar_flags = findPlanarCells(organized_array);
+#ifdef DEBUG_CAPE
+  planarCellsToLabels(planar_flags, "dbg_1_planar_cells.csv");
+  std::clog << "[DebugInfo] Planar cell found: " << planar_flags.count()
+            << '\n';
+#endif
+  // 2. Histogram initialization
+  Histogram hist = initializeHistogram(planar_flags);
+  // 3. Compute cell dist tols
+  std::vector<float> cell_dist_tols =
+      computeCellDistTols(organized_array, planar_flags);
+  // 4. Region growing
+  auto plane_segments = createPlaneSegments(hist, planar_flags, cell_dist_tols);
+#ifdef DEBUG_CAPE
+  planeSegmentsMapToLabels("dbg_2_plane_segments_raw.csv", _grid_plane_seg_map);
+  std::clog << "[DebugInfo] Plane segments found: " << plane_segments.size()
+            << '\n';
+#endif
+  // 5. Merge planes
+  std::vector<int32_t> merge_labels = mergePlanes(plane_segments);
+#ifdef DEBUG_CAPE
+  mergeSegmentsToLabels("dbg_3_plane_segments_merged.csv", merge_labels);
+  std::vector<int32_t> sorted_labels(merge_labels);
+  std::sort(sorted_labels.begin(), sorted_labels.end());
+
+  std::clog << "[DebugInfo] Planes number after merge: "
+            << std::distance(
+                   sorted_labels.begin(),
+                   std::unique(sorted_labels.begin(), sorted_labels.end()))
+            << '\n';
+#endif
+  // 6. Refinement (optional)
+  cv::Mat_ labels(_image_height, _image_width, 0);
+  if (_config.getBool("doRefinement")) {
+    refinePlanes(plane_segments, merge_labels, organized_array);
+    labels = toLabels();
+  } else {
+    labels = coarseToLabels(merge_labels);
+  }
+#ifdef DEBUG_CAPE
+  std::ofstream of("dbg_4_labels.csv");
+  of << cv::format(labels, cv::Formatter::FMT_CSV);
+#endif
+  // 7. Cleanup
+  cleanArtifacts();
+  return labels;
+}
+
+void CAPE::organizeByCell(Eigen::MatrixXf const& pcd_array,
```

Comment:
Please, specify in comment what form this array will have, I mean how data inside is organized